src/settings.c

```diff
@@ -275,62 +282,107 @@ static int settings_format_setting(struct setting *s, char *buf, int len)
   return buflen;
 }
 
-static void settings_msg_callback(u16 sender_id, u8 len, u8 msg[], void* context)
+static void settings_write_callback(u16 sender_id, u8 len, u8 msg[], void* context)
 {
-  (void)sender_id; (void) context;
+  (void) context;
+
+  if (sender_id != 0x42)
+    return;
 
   static struct setting *s = NULL;
   const char *section = NULL, *setting = NULL, *value = NULL;
-  char buf[256];
-  u8 buflen;
 
-  if (len == 0) {
-    /* Empty message is for parameter enumeration */
-    if (s && s->next)
-      s = s->next;
-    else
-      s = settings_head;
-  } else {
-    if (msg[len-1] != '\0')
-      goto error;
-    /* Extract parameters from message:
-     * 2 or 3 null terminated strings: section, setting and (optional) value
-     * If value is present the message is an assignment.
-     */
-    section = (const char *)msg;
-    for (int i = 0, tok = 0; i < len; i++) {
-      if (msg[i] == '\0') {
-        tok++;
-        switch (tok) {
-        case 1:
-          setting = (const char *)&msg[i+1];
-          break;
-        case 2:
-          if (i + 1 < len)
-            value = (const char *)&msg[i+1];
+  if (len == 0)
+    goto error;
```

Comment:
We've tried to avoid `goto` elsewhere, this function is the only place it's currently used - maybe just `log_error` and `return`?